app/src/main/java/com/github/polypoly/app/network/RemoteDB.kt

```diff
@@ -0,0 +1,131 @@
+            if (userRef.value == null) {
+                future.completeExceptionally(IllegalAccessError("No user with id $userId"))
+            } else {
+                future.complete(userRef.getValue<User>())
+            }
+        }.addOnFailureListener(future::completeExceptionally)
+
+        return future
+    }
+
+    override fun getAllUsers(): CompletableFuture<List<User>> {
+        val usersPromise = CompletableFuture<List<User>>()
+
+        usersRootRef.get().addOnSuccessListener{ usersRef ->
+            val users = ArrayList<User>()
+            for (child in usersRef.children) {
+                users.add(child.getValue<User>()!!)
+            }
+            usersPromise.complete(users)
+        }.addOnFailureListener(usersPromise::completeExceptionally)
+
+        return usersPromise
+    }
+
+    override fun getAllUsersIds(): CompletableFuture<List<Long>> {
+        val usersIdsPromise = CompletableFuture<List<Long>>()
+
+        usersRootRef.get().addOnSuccessListener{ usersRef ->
+            val ids = ArrayList<Long>()
+            for (child in usersRef.children) {
+                ids.add(child.key?.toLong() ?: -1)
+            }
+            usersIdsPromise.complete(ids)
+        }.addOnFailureListener(usersIdsPromise::completeExceptionally)
+
+        return usersIdsPromise
+    }
+
+    override fun registerUser(user: User): CompletableFuture<Boolean> {
+        return getAllUsersIds().thenCompose { allIds ->
+            val registerPromise = CompletableFuture<Boolean>()
+
+            if (allIds.contains(user.id)) {
+                registerPromise.completeExceptionally(
+                    IllegalAccessError("You cannot register two times a single account"))
+            } else {
+                usersRootRef.child(user.id.toString()).setValue(user).addOnSuccessListener {
+                    registerPromise.complete(true)
+                }.addOnFailureListener(registerPromise::completeExceptionally)
+            }
+
+            registerPromise
+        }
+    }
+
+    override fun updateUser(user: User): CompletableFuture<Boolean> {
+        return getAllUsersIds().thenCompose { allIds ->
+            val registerPromise = CompletableFuture<Boolean>()
+
+            if (!allIds.contains(user.id)) {
+                registerPromise.completeExceptionally(
+                    IllegalAccessError("Tries to update a user not registered yet"))
+            } else {
+                usersRootRef.updateChildren(mapOf(
+                    user.id.toString() to user
+                )).addOnSuccessListener {
+                    registerPromise.complete(true)
+                }.addOnFailureListener(registerPromise::completeExceptionally)
+            }
+
+            registerPromise
+        }
+    }
+
+    fun getUnderlyingDB(): FirebaseDatabase{
+        return db!!
+    }
+
+    companion object InvalidRemoteDB: RemoteDB(null, "") {
```

Comment:
FWIW, there is such a thing as the [null object pattern](https://en.wikipedia.org/wiki/Null_object_pattern), however this is not it: there's no point in throwing those exceptions compared to literally just using `null`. It would make sense if you intended to, e.g., return no users at all to represent an empty/unavailable DB.